File: OpenBurnBarCore/Sources/OpenBurnBarComputerUseCore/CapabilityTokenVerifier.swift
import Foundation
import CryptoKit

public enum CapabilityTokenVerificationFailure: String, Codable, Sendable, Equatable, Error {
    case missingToken = "capability_token_missing"
    case missingSignature = "capability_token_missing_signature"
    case signatureInvalid = "capability_token_signature_invalid"
    case expired = "capability_token_expired"
    case domainMismatch = "capability_token_domain_mismatch"
    case nonceReplay = "capability_token_nonce_replay"
    case actionNotAllowed = "capability_token_action_not_allowed"
    case actionBudgetExhausted = "capability_token_action_budget_exhausted"
    case attestationMismatch = "capability_token_attestation_mismatch"
    case escrowDeviceMismatch = "capability_token_escrow_device_mismatch"
    case issuerKeyUnavailable = "capability_token_issuer_key_unavailable"
    case issuerRevoked = "capability_token_issuer_revoked"
}

/// Tracks consumed single-use nonces for offline leaf verification.
public protocol CapabilityTokenNonceStore: Sendable {
    func hasConsumed(nonce: String, domain: CapabilityToken.Domain) -> Bool
    func consume(nonce: String, domain: CapabilityToken.Domain) throws
}

/// In-memory nonce store for unit tests and ephemeral coordinators.
public final class InMemoryCapabilityTokenNonceStore: CapabilityTokenNonceStore, @unchecked Sendable {
    private let lock = NSLock()
    private var consumed: Set<String> = []

    public init() {}

    private func key(nonce: String, domain: CapabilityToken.Domain) -> String {
        "\(domain.rawValue):\(nonce)"
    }

    public func hasConsumed(nonce: String, domain: CapabilityToken.Domain) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        return consumed.contains(key(nonce: nonce, domain: domain))
    }

    public func consume(nonce: String, domain: CapabilityToken.Domain) throws {
        lock.lock()
        defer { lock.unlock() }
        let entry = key(nonce: nonce, domain: domain)
        guard !consumed.contains(entry) else {
            throw CapabilityTokenVerificationFailure.nonceReplay
        }
        consumed.insert(entry)
    }
}

/// File-backed nonce ledger shared by the Virtual HID bridge (offline, no network).
public final class FileCapabilityTokenNonceStore: CapabilityTokenNonceStore, @unchecked Sendable {
    public struct Ledger: Codable, Sendable, Equatable {
        public var consumed: [String]
        public init(consumed: [String] = []) {
            self.consumed = consumed
        }
    }

    private let path: String
    private let lock = NSLock()
    private let maximumEntries = 10_000

    public init(path: String = RemoteUnlockSetupProbe.capabilityTokenNonceLedgerPath) {
        self.path = path
    }

    public func hasConsumed(nonce: String, domain: CapabilityToken.Domain) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        return loadLocked().consumed.contains(entryKey(nonce: nonce, domain: domain))
    }

    public func consume(nonce: String, domain: CapabilityToken.Domain) throws {
        lock.lock()
        defer { lock.unlock() }
        var ledger = loadLocked()
        let entry = entryKey(nonce: nonce, domain: domain)
        guard !ledger.consumed.contains(entry) else {
            throw CapabilityTokenVerificationFailure.nonceReplay
        }
        ledger.consumed.append(entry)
        if ledger.consumed.count > maximumEntries {
            ledger.consumed = Array(ledger.consumed.suffix(maximumEntries / 2))
        }
        try persistLocked(ledger)
    }

    private func entryKey(nonce: String, domain: CapabilityToken.Domain) -> String {
        "\(domain.rawValue):\(nonce)"
    }

    private func loadLocked() -> Ledger {
        guard let data = try? Data(contentsOf: URL(fileURLWithPath: path)),
              let ledger = try? JSONDecoder().decode(Ledger.self, from: data) else {
            return Ledger()
        }
        return ledger
    }

    private func persistLocked(_ ledger: Ledger) throws {
        let url = URL(fileURLWithPath: path)
        let directory = url.deletingLastPathComponent()
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        let data = try JSONEncoder().encode(ledger)
        try data.write(to: url, options: .atomic)
        try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: path)
    }
}

public struct CapabilityTokenIssuerTrust: Sendable {
    public var publicKey: Curve25519.Signing.PublicKey
    public var keyId: String
    public var revoked: Bool

    public init(publicKey: Curve25519.Signing.PublicKey, keyId: String, revoked: Bool = false) {
        self.publicKey = publicKey
        self.keyId = keyId
        self.revoked = revoked
    }
}

/// Offline leaf verifier: signature + TTL + nonce-unseen + action ∈ allowed (+ attestation binding).
public struct CapabilityTokenLeafVerifier: Sendable {
    public struct Request: Sendable, Equatable {
        public var token: CapabilityToken?
        public var expectedDomain: CapabilityToken.Domain
        public var actionKind: String
        public var requiredAttestationHashBlake3: String?
        public var boundEscrowDeviceId: String?

        public init(
            token: CapabilityToken?,
            expectedDomain: CapabilityToken.Domain,
            actionKind: String,
            requiredAttestationHashBlake3: String? = nil,
            boundEscrowDeviceId: String? = nil
        ) {
            self.token = token
            self.expectedDomain = expectedDomain
            self.actionKind = actionKind
            self.requiredAttestationHashBlake3 = requiredAttestationHashBlake3
            self.boundEscrowDeviceId = boundEscrowDeviceId
        }
    }

    private let signer = CapabilityTokenSigner()
    private let nonceStore: CapabilityTokenNonceStore
    private let issuerTrustProvider: @Sendable () -> CapabilityTokenIssuerTrust?

    public init(
        nonceStore: CapabilityTokenNonceStore,
        issuerTrust: @escaping @Sendable () -> CapabilityTokenIssuerTrust?
    ) {
        self.nonceStore = nonceStore
        self.issuerTrustProvider = issuerTrust
    }

    public func verify(request: Request, now: Date = Date()) -> Result<Void, CapabilityTokenVerificationFailure> {
        guard let token = request.token else {
            return .failure(.missingToken)
        }
        guard token.domain == request.expectedDomain else {
            return .failure(.domainMismatch)
        }
        guard !token.isExpired(at: now) else {
            return .failure(.expired)
        }
        guard token.signatureEd25519Base64 != nil else {
            return .failure(.missingSignature)
        }
        guard let trust = issuerTrustProvider() else {
            return .failure(.issuerKeyUnavailable)
        }
        guard !trust.revoked else {
            return .failure(.issuerRevoked)
        }
        guard (try? signer.verify(token: token, publicKey: trust.publicKey)) == true else {
            return .failure(.signatureInvalid)
        }
        if let required = request.requiredAttestationHashBlake3 {
            guard token.attestationHashBlake3 == required else {
                return .failure(.attestationMismatch)
            }
        }
        if let bound = request.boundEscrowDeviceId ?? token.boundEscrowDeviceId {
            if let tokenBound = token.boundEscrowDeviceId, tokenBound != bound {
                return .failure(.escrowDeviceMismatch)
            }
            if token.boundEscrowDeviceId == nil, request.boundEscrowDeviceId != nil {
                return .failure(.escrowDeviceMismatch)
            }
        }
        let normalizedKind = request.actionKind.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard token.allows(actionKind: normalizedKind) else {
            return .failure(.actionNotAllowed)
        }
        guard token.actionBudget > 0 else {
            return .failure(.actionBudgetExhausted)
        }
        if nonceStore.hasConsumed(nonce: token.nonce, domain: token.domain) {
            return .failure(.nonceReplay)
        }
        do {
            try nonceStore.consume(nonce: token.nonce, domain: token.domain)
        } catch let error as CapabilityTokenVerificationFailure {
            return .failure(error)
        } catch {
            return .failure(.nonceReplay)
        }
        return .success(())
    }
}

File: windows/computeruse/OpenBurnBar.ComputerUse.Core/Capability/CapabilityTokenVerifier.cs
// Offline leaf verifier for capability tokens.
//
// Port of CapabilityTokenVerifier.swift. Verifies, in the SAME order as the
// Swift core so the harshest failure wins:
//   missingToken -> domainMismatch -> expired -> missingSignature ->
//   issuerKeyUnavailable -> issuerRevoked -> signatureInvalid -> scopeMismatch ->
//   attestation binding (no self-matching) -> escrow-device binding ->
//   actionNotAllowed -> actionBudgetExhausted -> nonceReplay (check then consume).

using System;
using System.Collections.Generic;
using OpenBurnBar.ComputerUse.Core.Crypto;

namespace OpenBurnBar.ComputerUse.Core.Capability;

/// <summary>Why a capability-token verification failed. Wire strings match the Swift core.</summary>
public enum CapabilityTokenVerificationFailure
{
    None = 0,
    MissingToken,
    MissingSignature,
    SignatureInvalid,
    Expired,
    DomainMismatch,
    NonceReplay,
    ActionNotAllowed,
    ActionBudgetExhausted,
    ScopeMismatch,
    AttestationMismatch,
    EscrowDeviceMismatch,
    IssuerKeyUnavailable,
    IssuerRevoked,
}

/// <summary>Stable wire strings for <see cref="CapabilityTokenVerificationFailure"/>.</summary>
public static class CapabilityTokenVerificationFailureWire
{
    public static string ToWire(this CapabilityTokenVerificationFailure failure) => failure switch
    {
        CapabilityTokenVerificationFailure.MissingToken => "capability_token_missing",
        CapabilityTokenVerificationFailure.MissingSignature => "capability_token_missing_signature",
        CapabilityTokenVerificationFailure.SignatureInvalid => "capability_token_signature_invalid",
        CapabilityTokenVerificationFailure.Expired => "capability_token_expired",
        CapabilityTokenVerificationFailure.DomainMismatch => "capability_token_domain_mismatch",
        CapabilityTokenVerificationFailure.NonceReplay => "capability_token_nonce_replay",
        CapabilityTokenVerificationFailure.ActionNotAllowed => "capability_token_action_not_allowed",
        CapabilityTokenVerificationFailure.ActionBudgetExhausted => "capability_token_action_budget_exhausted",
        CapabilityTokenVerificationFailure.ScopeMismatch => "capability_token_scope_mismatch",
        CapabilityTokenVerificationFailure.AttestationMismatch => "capability_token_attestation_mismatch",
        CapabilityTokenVerificationFailure.EscrowDeviceMismatch => "capability_token_escrow_device_mismatch",
        CapabilityTokenVerificationFailure.IssuerKeyUnavailable => "capability_token_issuer_key_unavailable",
        CapabilityTokenVerificationFailure.IssuerRevoked => "capability_token_issuer_revoked",
        CapabilityTokenVerificationFailure.None => "none",
        _ => throw new ArgumentOutOfRangeException(nameof(failure), failure, null),
    };
}

/// <summary>Tracks consumed single-use nonces for offline leaf verification.</summary>
public interface ICapabilityTokenNonceStore
{
    /// <summary>True iff (<paramref name="nonce"/>, <paramref name="domain"/>) was consumed.</summary>
    bool HasConsumed(string nonce, CapabilityDomain domain);

    /// <summary>Marks (<paramref name="nonce"/>, <paramref name="domain"/>) consumed;
    /// throws <see cref="CapabilityTokenReplayException"/> if already consumed.</summary>
    void Consume(string nonce, CapabilityDomain domain);
}

/// <summary>Thrown when a nonce is consumed twice.</summary>
public sealed class CapabilityTokenReplayException : Exception
{
    public CapabilityTokenReplayException()
        : base("Capability-token nonce replay.")
    {
    }
}

/// <summary>In-memory nonce store for tests and ephemeral coordinators.</summary>
public sealed class InMemoryCapabilityTokenNonceStore : ICapabilityTokenNonceStore
{
    private readonly HashSet<string> _consumed = new(StringComparer.Ordinal);
    private readonly object _gate = new();

    public bool HasConsumed(string nonce, CapabilityDomain domain)
    {
        lock (_gate)
        {
            return _consumed.Contains(Key(nonce, domain));
        }
    }

    public void Consume(string nonce, CapabilityDomain domain)
    {
        lock (_gate)
        {
            if (!_consumed.Add(Key(nonce, domain)))
            {
                throw new CapabilityTokenReplayException();
            }
        }
    }

    private static string Key(string nonce, CapabilityDomain domain) => $"{domain.ToWire()}:{nonce}";
}

/// <summary>Published issuer trust: the pinned public key + revocation flag.</summary>
public sealed class CapabilityTokenIssuerTrust
{
    public CapabilityTokenIssuerTrust(ICapabilityVerifier verifier, string keyId, bool revoked = false)
    {
        Verifier = verifier ?? throw new ArgumentNullException(nameof(verifier));
        KeyId = keyId ?? throw new ArgumentNullException(nameof(keyId));
        Revoked = revoked;
    }

    public ICapabilityVerifier Verifier { get; }

    public string KeyId { get; }

    public bool Revoked { get; }
}

/// <summary>A verification request: the token plus the bindings the presenter asserts.</summary>
public sealed class CapabilityTokenVerificationRequest
{
    public CapabilityTokenVerificationRequest(
        CapabilityToken? token,
        CapabilityDomain expectedDomain,
        string actionKind,
        string? requiredScopeHash = null,
        string? requiredAttestationHashBlake3 = null,
        string? boundEscrowDeviceId = null)
    {
        Token = token;
        ExpectedDomain = expectedDomain;
        ActionKind = actionKind ?? throw new ArgumentNullException(nameof(actionKind));
        RequiredScopeHash = Normalize(requiredScopeHash);
        RequiredAttestationHashBlake3 = Normalize(requiredAttestationHashBlake3);
        BoundEscrowDeviceId = Normalize(boundEscrowDeviceId);
    }

    public CapabilityToken? Token { get; }

    public CapabilityDomain ExpectedDomain { get; }

    public string ActionKind { get; }

    public string? RequiredScopeHash { get; }

    public string? RequiredAttestationHashBlake3 { get; }

    public string? BoundEscrowDeviceId { get; }

    private static string? Normalize(string? value)
    {
        if (value is null)
        {
            return null;
        }

        var trimmed = value.Trim();
        return trimmed.Length == 0 ? null : trimmed;
    }
}

/// <summary>The verdict of a leaf verification: success, or the harshest failure.</summary>
public readonly struct CapabilityTokenVerificationResult
{
    private CapabilityTokenVerificationResult(bool ok, CapabilityTokenVerificationFailure failure)
    {
        IsSuccess = ok;
        Failure = failure;
    }

    public bool IsSuccess { get; }

    public CapabilityTokenVerificationFailure Failure { get; }

    public static CapabilityTokenVerificationResult Success { get; } =
        new(true, CapabilityTokenVerificationFailure.None);

    public static CapabilityTokenVerificationResult Fail(CapabilityTokenVerificationFailure failure)
        => new(false, failure);
}

/// <summary>
/// Offline leaf verifier: signature + TTL + domain + scope/attestation/escrow
/// bindings + action-allowed + budget + single-use nonce. Attestation and
/// escrow bindings never self-match: if the token is bound, the presenter must
/// supply the expected binding from a trusted context.
/// </summary>
public sealed class CapabilityTokenLeafVerifier
{
    private readonly CapabilityTokenSigner _signer = new();
    private readonly ICapabilityTokenNonceStore _nonceStore;
    private readonly Func<CapabilityTokenIssuerTrust?> _issuerTrustProvider;

    public CapabilityTokenLeafVerifier(
        ICapabilityTokenNonceStore nonceStore,
        Func<CapabilityTokenIssuerTrust?> issuerTrustProvider)
    {
        _nonceStore = nonceStore ?? throw new ArgumentNullException(nameof(nonceStore));
        _issuerTrustProvider = issuerTrustProvider ?? throw new ArgumentNullException(nameof(issuerTrustProvider));
    }

    public CapabilityTokenVerificationResult Verify(
        CapabilityTokenVerificationRequest request,
        DateTimeOffset now)
    {
        var token = request.Token;
        if (token is null)
        {
            return Fail(CapabilityTokenVerificationFailure.MissingToken);
        }

        if (token.Domain != request.ExpectedDomain)
        {
            return Fail(CapabilityTokenVerificationFailure.DomainMismatch);
        }

        if (token.IsExpired(now))
        {
            return Fail(CapabilityTokenVerificationFailure.Expired);
        }

        if (string.IsNullOrEmpty(token.SignatureEd25519Base64))
        {
            return Fail(CapabilityTokenVerificationFailure.MissingSignature);
        }

        var trust = _issuerTrustProvider();
        if (trust is null)
        {
            return Fail(CapabilityTokenVerificationFailure.IssuerKeyUnavailable);
        }

        if (trust.Revoked)
        {
            return Fail(CapabilityTokenVerificationFailure.IssuerRevoked);
        }

        if (!_signer.Verify(token, trust.Verifier))
        {
            return Fail(CapabilityTokenVerificationFailure.SignatureInvalid);
        }

        if (request.RequiredScopeHash is not null && token.ScopeHash != request.RequiredScopeHash)
        {
            return Fail(CapabilityTokenVerificationFailure.ScopeMismatch);
        }

        // Attestation binding without self-matching: if either side is bound,
        // both must be present and equal.
        var tokenAttestation = token.AttestationHashBlake3;
        if (request.RequiredAttestationHashBlake3 is not null || tokenAttestation is not null)
        {
            if (request.RequiredAttestationHashBlake3 is null
                || tokenAttestation is null
                || tokenAttestation != request.RequiredAttestationHashBlake3)
            {
                return Fail(CapabilityTokenVerificationFailure.AttestationMismatch);
            }
        }

        // Escrow-device binding follows the same rule.
        var tokenDevice = token.BoundEscrowDeviceId;
        if (request.BoundEscrowDeviceId is not null || tokenDevice is not null)
        {
            if (request.BoundEscrowDeviceId is null
                || tokenDevice is null
                || tokenDevice != request.BoundEscrowDeviceId)
            {
                return Fail(CapabilityTokenVerificationFailure.EscrowDeviceMismatch);
            }
        }

        var normalizedKind = request.ActionKind.Trim();
        if (!token.Allows(normalizedKind))
        {
            return Fail(CapabilityTokenVerificationFailure.ActionNotAllowed);
        }

        if (token.ActionBudget <= 0)
        {
            return Fail(CapabilityTokenVerificationFailure.ActionBudgetExhausted);
        }

        if (_nonceStore.HasConsumed(token.Nonce, token.Domain))
        {
            return Fail(CapabilityTokenVerificationFailure.NonceReplay);
        }

        try
        {
            _nonceStore.Consume(token.Nonce, token.Domain);
        }
        catch (CapabilityTokenReplayException)
        {
            return Fail(CapabilityTokenVerificationFailure.NonceReplay);
        }

        return CapabilityTokenVerificationResult.Success;
    }

    private static CapabilityTokenVerificationResult Fail(CapabilityTokenVerificationFailure failure)
        => CapabilityTokenVerificationResult.Fail(failure);
}
